### PR TITLE
[FEATURE] Add functionality to manage partial project deliverables

### LINKED ISSUE #9: [FEATURE] Add functionality to manage partial project deliverables.
**Is your feature request related to a problem? Please describe.**
Projects should have partial project deliverables which could be defined, edited and removed by team members. Students should be able to update a deliverable with the link to the finished product.

**Describe the solution you'd like**
Create a router to expose the endpoints needed to manage partial project deliverables.

**Describe alternatives you've considered**
N/A

**Provide usage examples**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by b0tw**
Closes #9

### CODE REVIEW

File: server/routers/deliverables.js [hunk 1]
```diff
@@ -14,55 +17,66 @@ router.post('/add', (req, res) => {
...
+            checkDeliverable.title = title
+        }
+        if (description !== null && description.length < 200) {
+            checkDeliverable.description = description
+        }
+        if (link !== null) {
+            checkDeliverable.link = link
+        }
+
+        checkDeliverable.save()
```
Comment (filipdutescu) [inline server/routers/deliverables.js]: I think this is async

**Review by filipdutescu — APPROVED** [review]

**Comment by filipdutescu** [comment]
Did you also rebase? I doesn't tell me.

**Review by MausoleumREST — APPROVED** [review]

File: server/routers/deliverables.js [hunk 2]
```diff
@@ -14,55 +17,66 @@ router.post('/add', (req, res) => {
...
-    //response (will be a message after the db is made)
-    res.status(200).send(deliverable)
 })
 
-router.get('/get', (req, res) => {
-    projectId = req.body.projectId
-
-    //search in db
+router.get('/get', async (req, res) => {
+    id = req.body.id
```
Comment (filipdutescu) [inline server/routers/deliverables.js]: No body in GET. Please also preserve the order of the request, as you defined in your previous PR

**Review by filipdutescu — CHANGES_REQUESTED** [review]

File: server/routers/deliverables.js [hunk 3]
```diff
@@ -22,52 +28,63 @@ router.post('/add', (req, res) => {
...
+    }
+    else {
+        res.status(400).send({ "message": "Deliverable or project not found" })
+    }
 })
 
-router.put('/update', (req, res) => {
-    projectId = req.body.projectId
+router.put('/update', async (req, res) => {
+    id = req.body.id
```
Comment (filipdutescu) [inline server/routers/deliverables.js]:
```suggestion
router.put('/:id', async (req, res) => {
    let id = req.params.id,
```

File: server/routers/deliverables.js [hunk 4]
```diff
@@ -22,52 +28,63 @@ router.post('/add', (req, res) => {
...
-    //response (will be a message after the db is made)
-    res.status(200).send(deliverable)
 })
 
-router.get('/get', (req, res) => {
-    projectId = req.body.projectId
-
-    //search in db
-
+router.get('/get', async (req, res) => {
```
Comment (filipdutescu) [inline server/routers/deliverables.js]:
```suggestion
router.get('/', async (req, res) => {
```

File: server/routers/deliverables.js [hunk 5]
```diff
@@ -22,52 +28,63 @@ router.post('/add', (req, res) => {
...
+    id = req.body.id
```
Comment (filipdutescu) [inline server/routers/deliverables.js]:
```suggestion
    let id = req.query.id
```

File: server/routers/deliverables.js [hunk 6]
```diff
@@ -1,16 +1,22 @@
...
 const authMiddleware = require('../middleware/authentication-middleware');
 const apiError = require('../entities/api-error');
 
 // TODO: uncomment and test after session router is made
 router.use(authMiddleware());
 
-router.post('/add', (req, res) => {
-    projectId = req.body.projectId
+
+router.post('/add', async (req, res) => {
```
Comment (filipdutescu) [inline server/routers/deliverables.js]:
```suggestion
router.post('/', async (req, res) => {
```

File: server/routers/deliverables.js [hunk 7]
```diff
@@ -1,16 +1,22 @@
...
 const router = express.Router();
+
+const context = require('../entities/database/context')
+const Deliverable = context.Deliverable
+const Project = context.Project
+
 const authMiddleware = require('../middleware/authentication-middleware');
 const apiError = require('../entities/api-error');
 
 // TODO: uncomment and test after session router is made
```
Comment (filipdutescu) [inline server/routers/deliverables.js]:
```suggestion
```

**Review by filipdutescu — CHANGES_REQUESTED** [review]

**Review by MausoleumREST — APPROVED** [review]

File: server/routers/deliverables.js (hunk 7 above)
Comment (b0tw) [reply server/routers/deliverables.js]: Cannot easily test that right now as there is no route to register users and the password field in the database should be a hash

File: server/routers/deliverables.js [hunk 8]
```diff
@@ -5,12 +5,28 @@ const apiError = require('../entities/api-error');
...
+
+    if (deliverable !== null) {
+        res.status(200).send({ deliverable })
+    }
+    else {
+        res.status(400).send({ "message": "Deliverable or project not found" })
+    }
+})
+
+router.post('/add', async (req, res) => {
```
Comment (filipdutescu) [inline server/routers/deliverables.js]:
```suggestion
router.post('/', async (req, res) => {
```

File: server/routers/deliverables.js [hunk 9]
```diff
@@ -19,55 +35,62 @@ router.post('/add', (req, res) => {
     else if (description.length > 200) {
         res.status(400).send(apiError.InvalidRequest);
     }
-    else if(link ===null){
+    else if(url ===null){
```
Comment (filipdutescu) [inline server/routers/deliverables.js]:
```suggestion
    else if(url == null){
```

File: server/routers/deliverables.js [hunk 10]
```diff
@@ -19,55 +35,62 @@ router.post('/add', (req, res) => {
...
+    try {
+        const del = await Deliverable.create({ title, description, url, project_id })
 
-    res.status(200).send(projectId)
+        res.status(200).send({ "message": del.title + " was created." })
+    }
+    catch (err) {
+        console.log(err)
+        res.status(400).send({ "message": "Something bad happened" })
+    }
```
Comment (filipdutescu) [inline server/routers/deliverables.js]: this seems to be duplicated

File: server/routers/deliverables.js [hunk 11]
```diff
@@ -19,55 +35,62 @@ router.post('/add', (req, res) => {
...
     description = req.body.description
     link = req.body.link
-    //validate projectId
-
-
-    if (title !== null && title.length < 10) {
-        //update title in the db
+    project_id = req.body.project_id
 
+    let checkDeliverable = await Deliverable.findOne({ where: { id, project_id } })
```
Comment (filipdutescu) [inline server/routers/deliverables.js:64]: honestly can't believe this works and you don't need to specify `{ id: id, project_id: project_id }`

**Review by filipdutescu — CHANGES_REQUESTED** [review]
let's fix the few comments and then we can merge. Solid work!

File: server/routers/deliverables.js (hunk 11 above)
Comment (b0tw) [reply server/routers/deliverables.js:64]: If the variables have the variable has the same name with the property, it works. It's way easier to write like this.

File: server/appsettings.json [hunk 12]
```diff
@@ -3,8 +3,8 @@
   "connections": {
     "database": {
       "name": "anon_grading_db",
-      "username": "sequelize",
-      "password": "_?aaD045.h",
+      "username": "root",
+      "password": "",
```
Comment (filipdutescu) [inline server/appsettings.json]: revert this please. This is a security issue. We should never connect as a DBA through an ORM

**Review by filipdutescu — CHANGES_REQUESTED** [review]

**Review by filipdutescu — APPROVED** [review]

**Pull request merged**